### ISSUE TITLE
libnewrelic.so is not linked against libpcre

### ISSUE THREAD

**Issue by burgerrg** [newrelic/c-sdk#9]
The libnewrelic.so shared object is not linked against libpcre.so (as shown by ldd), so that when it's loaded into a program such as Chez Scheme, it throws this error:
undefined_symbol: pcre_free

If you add -lpcre to the end of the line that builds the .so, it will build correctly in Linux.

On macOS, you can first extract the members of libnewrelic.a and then build libnewrelic.dylib.

Here's one way to change the Makefile to do this (and you have my permission to use this change):

```
diff --git a/.gitignore b/.gitignore
index 88d356d..f82297d 100644
--- a/.gitignore
+++ b/.gitignore
@@ -36,6 +36,7 @@ newrelic-daemon
 #
 # Byproducts from make libnewrelic.so
 #
+libnewrelic.dylib
 libnewrelic.so
 
 #
diff --git a/Makefile b/Makefile
index 7fe575d..dc8c472 100644
--- a/Makefile
+++ b/Makefile
@@ -29,6 +29,7 @@ export CMOCKA_INCLUDE
 # OS X 10.11 does not provide pcre-config by default.
 # Check whether it exists, and if not assume a sensible default.
 PCRE_CFLAGS := $(shell pcre-config --cflags)
+PCRE_LIBS := $(shell pcre-config --libs)
 
 #
 # We pull in the current SDK version from the VERSION file, and expose it to
@@ -47,11 +48,13 @@ ifeq (Darwin,$(UNAME))
 # then uses macOS's special libtool to smoosh them together into a single,
 # beautiful library.
 #
+SONAME := libnewrelic.dylib
 LIBTOOL := /usr/bin/libtool
 
 libnewrelic.a: axiom src-static
 	$(LIBTOOL) -static -o $@ vendor/newrelic/axiom/libaxiom.a src/libnewrelic.a
 else
+SONAME := libnewrelic.so
 #
 # This rule builds a static axiom library and a static C SDK library, and
 # then uses GNU ar's MRI support to smoosh them together into a single,
@@ -102,7 +105,7 @@ daemon-clean:
 	rm -f newrelic-daemon
 
 .PHONY: dynamic
-dynamic: libnewrelic.so
+dynamic: $(SONAME)
 
 .PHONY: valgrind
 valgrind: vendor libnewrelic.a
@@ -113,7 +116,12 @@ valgrind: vendor libnewrelic.a
 # the static library and have gcc wrap it in the appropriate shared library
 # goop.
 libnewrelic.so: libnewrelic.a
-	$(CC) -shared -pthread $(PCRE_CFLAGS) -ldl -o $@ -Wl,--whole-archive $^  -Wl,--no-whole-archive
+	$(CC) -shared -pthread -ldl -o $@ -Wl,--whole-archive $^ -Wl,--no-whole-archive $(PCRE_LIBS)
+
+libnewrelic.dylib: libnewrelic.a
+	$(AR) -x $<
+	$(CC) -shared -o $@ *.o $(PCRE_LIBS)
+	rm -f *.o
 
 .PHONY: src-static
 src-static:
@@ -130,7 +138,7 @@ tests-clean:
 
 .PHONY: clean
 clean: axiom-clean daemon-clean src-clean tests-clean
-	rm -f *.o libnewrelic.a libnewrelic.so
+	rm -f *.o libnewrelic.a libnewrelic.dylib libnewrelic.so
 
 .PHONY: integration
 integration: libnewrelic.so daemon
```

**Comment by pyohannes** [comment]
Hi Bob! Thanks very much for reporting this bug. We will add the PCRE dependency to `libnewrelic.so`. The fix will ship with the next C SDK release. We won't add support for creating a `.dylib` library as we don't officially support MacOS.

I'm excited to learn that you're using the C SDK with Chez Scheme, and I would love to learn more about that!

**Comment by burgerrg** [comment]
Thank you! The Mac's Xcode compiler doesn't support the --whole-archive flag. If you update the Makefile's libnewrelic.so target to use $(AR) -x followed by $(CC) -shared as in the libnewrelic.dylib target above, it will work on both Linux and Mac.

**Comment by burgerrg** [comment]
My company uses [Chez Scheme](https://github.com/cisco/ChezScheme) and the [Swish Concurrency Engine](https://github.com/becls/swish) to provide services through a simple HTTP server. The libnewrelic.so shared object allows me to use the New Relic API in Scheme without having to write any C code.

**Comment by burgerrg** [comment]
Here's what I had in mind:
```
diff --git a/Makefile b/Makefile
index 7fe575d..9719fdb 100644
--- a/Makefile
+++ b/Makefile
@@ -29,6 +29,7 @@ export CMOCKA_INCLUDE
 # OS X 10.11 does not provide pcre-config by default.
 # Check whether it exists, and if not assume a sensible default.
 PCRE_CFLAGS := $(shell pcre-config --cflags)
+PCRE_LIBS := $(shell pcre-config --libs)
 
 #
 # We pull in the current SDK version from the VERSION file, and expose it to
@@ -113,7 +114,9 @@ valgrind: vendor libnewrelic.a
 # the static library and have gcc wrap it in the appropriate shared library
 # goop.
 libnewrelic.so: libnewrelic.a
-	$(CC) -shared -pthread $(PCRE_CFLAGS) -ldl -o $@ -Wl,--whole-archive $^  -Wl,--no-whole-archive
+	$(AR) -x $<
+	$(CC) -shared -pthread -ldl -o $@ *.o $(PCRE_LIBS)
+	rm -f *.o
 
 .PHONY: src-static
 src-static:
```

**Comment by Fahmy-Mohammed** [comment]
Implemented in [5c62b25](https://github.com/newrelic/c-sdk/commit/5c62b253d6cbec199386873a89fa5542d46a6c81) and released in C-SDK version 1.2.0. Thank you for your feedback, we heard you!
